packages/backend/routes/users.ts

```diff
@@ -452,6 +487,190 @@ export default function userRoutes(app: Application) {
     }
   })
 
+  app.post('/api/login/mfa', [loginRateLimiter, optionalAuthentication], async (req: AuthorizedRequest, res) => {
+    let success = false
+    try {
+      if (req.body?.token && req.jwtData?.mfaStep == 1 && req.jwtData?.email) {
+        const userWithEmail = await User.findOne({
+          where: {
+            email: req.jwtData?.email,
+            banned: {
+              [Op.ne]: true
+            }
+          }
+        })
+        if (userWithEmail) {
+          const mfaDetails = await MfaDetails.findAll({
+            where: {
+              userId: userWithEmail.id,
+              enabled: {
+                [Op.eq]: true
+              }
+            }
+          })
+
+          let mfaPassed = false
+
+          for (let mfaDetail of mfaDetails) {
+            if (await verifyTotp(mfaDetail, req.body?.token)) {
+              mfaPassed = true
+              break
+            }
+          }
+
+          if (mfaPassed) {
+            success = true
+            res.send({
+              success: true,
+              token: jwt.sign(
+                {
+                  userId: userWithEmail.id,
+                  email: userWithEmail.email.toLowerCase(),
+                  birthDate: userWithEmail.birthDate,
+                  url: userWithEmail.url,
+                  role: userWithEmail.role
+                },
+                environment.jwtSecret,
+                { expiresIn: '31536000s' }
+              )
+            })
+            userWithEmail.lastLoginIp = getIp(req, true)
+            await userWithEmail.save()
+          }
+        }
+      }
+    } catch (error) {
+      logger.error(error)
+    }
+
+    if (!success) {
+      // res.statusCode = 401;
+      res.send({
+        success: false,
+        errorMessage: 'Invalid code provided'
+      })
+    }
+  })
+
+  // list all registered MFA options for a user
+  app.get('/api/user/mfa', authenticateToken, async (req: AuthorizedRequest, res) => {
+    if (req.jwtData?.userId) {
+      try {
+        const mfaDetails = await MfaDetails.findAll({
+          where: {
+            userId: req.jwtData?.userId,
+            enabled: {
+              [Op.eq]: true
+            }
+          }
+        })
+        res.send({
+          success: true,
+          mfa: mfaDetails.map((detail) => ({
+            id: detail.id,
+            name: detail.name,
+            type: detail.type,
+            enabled: detail.enabled
+          }))
+        })
+        return
+      } catch (error) {
+        logger.error(error)
+      }
+    }
+    res.send({ success: false })
+  })
+
+  app.post('/api/user/mfa', authenticateToken, async (req: AuthorizedRequest, res) => {
+    try {
+      if (req.jwtData?.userId && req.body?.type == "totp") {
```

Comment:
Do we have any other type that is not `totp`? Should we require this check now ?

Reply:
Just wanted to be defensive a bit, as the interface is only defined for totp right now anyway. Email and bypass would be the same interface though, but passkeys would need a slightly different one.